File: src/main/services/versionService.ts
import * as https from 'https'
import { getSettings } from './settingsService'
import { existsSync, readdirSync } from 'fs'
import { rm } from 'fs/promises'
import { join } from 'path'
import {
  downloadVersionJsonAndJar,
  resolveVersion,
  getCurrentPlatform,
  resolveLibraries,
  downloadLibraries,
  downloadAssetIndex,
  downloadAssets,
  linkLegacyAssets,
  fetchFabricLoaderArtifact,
  installFabricVersionJson,
  fetchQuiltLoaderArtifact,
  installQuiltVersionJson,
  installForge,
  installNeoForge,
} from './mcinstall'

export interface RemoteVersion {
  id: string
  type: 'release' | 'snapshot' | 'old_beta' | 'old_alpha'
  url: string
  time: string
  releaseTime: string
}

export interface VersionManifest {
  latest: { release: string; snapshot: string }
  versions: RemoteVersion[]
}

export interface FabricLoaderVersion {
  loader: { version: string; stable: boolean }
  intermediary: { version: string }
  launcherMeta: { version: number }
}

function httpsGet(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    https
      .get(url, res => {
        let data = ''
        res.on('data', c => (data += c))
        res.on('end', () => resolve(data))
      })
      .on('error', reject)
  })
}

// Fabric/Mojang APIs return plain-text error bodies (e.g. "no loader version X for
// game version Y") instead of JSON when a version/loader combo is invalid. JSON.parse
// on that text throws a cryptic "Unexpected token" SyntaxError — surface the actual
// response body instead so the failure is readable in the launch error dialog.
function safeJsonParse<T>(raw: string, context: string): T {
  try {
    return JSON.parse(raw) as T
  } catch {
    throw new Error(`${context}: ${raw.slice(0, 200).trim() || '(empty response)'}`)
  }
}

export async function fetchVersionManifest(): Promise<VersionManifest> {
  const raw = await httpsGet(
    'https://piston-meta.mojang.com/mc/game/version_manifest_v2.json',
  )
  return safeJsonParse<VersionManifest>(raw, 'Failed to fetch Minecraft version list')
}

export async function listFabricVersions(mcVersion: string): Promise<FabricLoaderVersion[]> {
  const raw = await httpsGet(
    `https://meta.fabricmc.net/v2/versions/loader/${mcVersion}`,
  )
  return safeJsonParse<FabricLoaderVersion[]>(raw, `No Fabric loader available for Minecraft ${mcVersion}`)
}

export async function listForgeVersions(mcVersion: string): Promise<string[]> {
  try {
    const raw = await httpsGet(
      'https://files.minecraftforge.net/net/minecraftforge/forge/promotions_slim.json',
    )
    const data = JSON.parse(raw)
    const promos: Record<string, string> = data.promos ?? {}
    const versions: string[] = []
    for (const [key, value] of Object.entries(promos)) {
      if (key.startsWith(`${mcVersion}-`)) {
        versions.push(String(value))
      }
    }
    return versions
  } catch {
    return []
  }
}

export function getInstalledVersions(): string[] {
  const settings = getSettings()
  const versionsDir = join(settings.game.defaultGameDir, 'versions')
  if (!existsSync(versionsDir)) return []
  try {
    return readdirSync(versionsDir, { withFileTypes: true })
      .filter(d => d.isDirectory())
      .map(d => d.name)
  } catch {
    return []
  }
}

export async function installVersion(
  versionId: string,
  loaderType: string,
  loaderVersion: string | undefined,
  onProgress: (task: string, progress: number, total: number) => void,
  javaPath = 'java',
  customGameDir?: string,
): Promise<void> {
  const settings = getSettings()
  const gameDir = customGameDir || settings.game.defaultGameDir
  const concurrency = settings.launcher.concurrentDownloads

  onProgress(`Installing Minecraft ${versionId}...`, 0, 100)

  const manifest = await fetchVersionManifest()
  const versionEntry = manifest.versions.find(v => v.id === versionId)
  if (!versionEntry) throw new Error(`Version ${versionId} not found in manifest`)

  await downloadVersionJsonAndJar(versionEntry, gameDir)
  const resolved = resolveVersion(gameDir, versionId)

  onProgress('Installing assets and libraries...', 50, 100)
  const platform = getCurrentPlatform()
  const libraries = resolveLibraries(resolved.libraries, platform)
  await downloadLibraries(libraries, join(gameDir, 'libraries'), concurrency)

  const assetIndex = await downloadAssetIndex(resolved, gameDir)
  await downloadAssets(assetIndex, gameDir, concurrency)
  await linkLegacyAssets(assetIndex, resolved, gameDir)

  if (loaderType === 'fabric' && loaderVersion) {
    onProgress(`Installing Fabric ${loaderVersion}...`, 80, 100)
    const artifact = await fetchFabricLoaderArtifact(versionId, loaderVersion)
    await installFabricVersionJson(artifact, versionId, gameDir)
  } else if (loaderType === 'quilt' && loaderVersion) {
    onProgress(`Installing Quilt ${loaderVersion}...`, 80, 100)
    const artifact = await fetchQuiltLoaderArtifact(versionId, loaderVersion)
    await installQuiltVersionJson(artifact, versionId, gameDir)
  } else if (loaderType === 'forge' && loaderVersion) {
    onProgress(`Installing Forge ${loaderVersion}...`, 80, 100)
    await installForge(versionId, loaderVersion, gameDir, javaPath, line => onProgress(line, 85, 100))
  } else if (loaderType === 'neoforge' && loaderVersion) {
    onProgress(`Installing NeoForge ${loaderVersion}...`, 80, 100)
    await installNeoForge(versionId, loaderVersion, gameDir, javaPath, line => onProgress(line, 85, 100))
  }

  onProgress('Done', 100, 100)
}

export async function deleteVersion(versionId: string): Promise<void> {
  const settings = getSettings()
  const versionDir = join(settings.game.defaultGameDir, 'versions', versionId)
  if (existsSync(versionDir)) {
    await rm(versionDir, { recursive: true, force: true })
  }
}

File: src-tauri/src/services/version_service.rs
//! Port of `src/main/services/versionService.ts`.

use crate::mcinstall::{assets, fabric, forge_install, libraries, platform, quilt, version_resolve};
use serde::{Deserialize, Serialize};
use std::path::Path;
use tokio::fs;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RemoteVersion {
    pub id: String,
    pub r#type: String,
    pub url: String,
    pub time: String,
    #[serde(rename = "releaseTime")]
    pub release_time: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct VersionManifestLatest {
    pub release: String,
    pub snapshot: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct VersionManifest {
    pub latest: VersionManifestLatest,
    pub versions: Vec<RemoteVersion>,
}

pub async fn fetch_version_manifest(client: &reqwest::Client) -> Result<VersionManifest, String> {
    let text = client
        .get("https://piston-meta.mojang.com/mc/game/version_manifest_v2.json")
        .send()
        .await
        .map_err(|e| e.to_string())?
        .text()
        .await
        .map_err(|e| e.to_string())?;
    serde_json::from_str(&text).map_err(|_| {
        format!("Failed to fetch Minecraft version list: {}", text.chars().take(200).collect::<String>())
    })
}

pub async fn list_forge_versions(client: &reqwest::Client, mc_version: &str) -> Vec<String> {
    let Ok(res) = client.get("https://files.minecraftforge.net/net/minecraftforge/forge/promotions_slim.json").send().await else {
        return Vec::new();
    };
    let Ok(data) = res.json::<serde_json::Value>().await else { return Vec::new() };
    let Some(promos) = data["promos"].as_object() else { return Vec::new() };
    let prefix = format!("{mc_version}-");
    promos
        .iter()
        .filter(|(k, _)| k.starts_with(&prefix))
        .filter_map(|(_, v)| v.as_str().map(|s| s.to_string()))
        .collect()
}

pub fn get_installed_versions(game_dir: &Path) -> Vec<String> {
    let versions_dir = game_dir.join("versions");
    let Ok(entries) = std::fs::read_dir(&versions_dir) else { return Vec::new() };
    entries
        .flatten()
        .filter(|e| e.path().is_dir())
        .filter_map(|e| e.file_name().into_string().ok())
        .collect()
}

pub async fn install_version(
    client: &reqwest::Client,
    version_id: &str,
    loader_type: &str,
    loader_version: Option<&str>,
    on_progress: impl Fn(String, u32, u32),
    game_dir: &Path,
    concurrency: usize,
    java_path: &str,
) -> Result<(), String> {
    on_progress(format!("Installing Minecraft {version_id}..."), 0, 100);

    let manifest = fetch_version_manifest(client).await?;
    let entry = manifest
        .versions
        .iter()
        .find(|v| v.id == version_id)
        .ok_or_else(|| format!("Version {version_id} not found in manifest"))?;

    version_resolve::download_version_json_and_jar(client, version_id, &entry.url, game_dir).await?;
    let resolved = version_resolve::resolve_version(game_dir, version_id).await?;

    on_progress("Installing assets and libraries...".to_string(), 50, 100);
    let plat = platform::get_current_platform();
    let libs = libraries::resolve_libraries(&resolved.libraries, &plat);
    libraries::download_libraries(client, libs, &game_dir.join("libraries"), concurrency).await?;

    let asset_index = assets::download_asset_index(client, &resolved, game_dir).await?;
    assets::download_assets(client, &asset_index, game_dir, concurrency).await?;
    assets::link_legacy_assets(&asset_index, &resolved, game_dir).await?;

    match (loader_type, loader_version) {
        ("fabric", Some(lv)) => {
            on_progress(format!("Installing Fabric {lv}..."), 80, 100);
            let artifact = fabric::fetch_fabric_loader_artifact(client, version_id, lv).await?;
            fabric::install_fabric_version_json(&artifact, version_id, game_dir).await?;
        }
        ("quilt", Some(lv)) => {
            on_progress(format!("Installing Quilt {lv}..."), 80, 100);
            let artifact = quilt::fetch_quilt_loader_artifact(client, version_id, lv).await?;
            quilt::install_quilt_version_json(&artifact, version_id, game_dir).await?;
        }
        ("forge", Some(lv)) => {
            on_progress(format!("Installing Forge {lv}..."), 80, 100);
            let on_progress_log = |line: String| on_progress(line, 85, 100);
            forge_install::install_forge(client, version_id, lv, game_dir, java_path, &on_progress_log).await?;
        }
        ("neoforge", Some(lv)) => {
            on_progress(format!("Installing NeoForge {lv}..."), 80, 100);
            let on_progress_log = |line: String| on_progress(line, 85, 100);
            forge_install::install_neoforge(client, lv, game_dir, java_path, &on_progress_log).await?;
        }
        _ => {}
    }

    on_progress("Done".to_string(), 100, 100);
    Ok(())
}

pub async fn delete_version(version_id: &str, game_dir: &Path) -> Result<(), String> {
    let dir = game_dir.join("versions").join(version_id);
    if fs::metadata(&dir).await.is_ok() {
        fs::remove_dir_all(&dir).await.map_err(|e| e.to_string())?;
    }
    Ok(())
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct FabricLoaderVersionEntry {
    pub loader: FabricLoaderVersionInfo,
    pub intermediary: FabricIntermediaryInfo,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct FabricLoaderVersionInfo {
    pub version: String,
    pub stable: bool,
}
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct FabricIntermediaryInfo {
    pub version: String,
}

pub async fn list_fabric_versions(client: &reqwest::Client, mc_version: &str) -> Result<Vec<FabricLoaderVersionEntry>, String> {
    let text = client
        .get(format!("https://meta.fabricmc.net/v2/versions/loader/{mc_version}"))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .text()
        .await
        .map_err(|e| e.to_string())?;
    serde_json::from_str(&text).map_err(|_| format!("No Fabric loader available for Minecraft {mc_version}"))
}
